google/resource_google_folder.go

```diff
@@ -0,0 +1,160 @@
+	config := meta.(*Config)
+
+	displayName := d.Get("display_name").(string)
+	parent := d.Get("parent").(string)
+
+	op, err := config.clientResourceManagerV2Beta1.Folders.Create(&resourceManagerV2Beta1.Folder{
+		DisplayName: displayName,
+	}).Parent(parent).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error creating folder '%s' in '%s': %s", displayName, parent, err)
+	}
+
+	err = resourceManagerV2Beta1OperationWait(config, op, "creating folder")
+
+	if err != nil {
+		return fmt.Errorf("Error creating folder '%s' in '%s': %s", displayName, parent, err)
+	}
+
+	// Since we waited above, the operation is guaranteed to have been successful by this point.
+	waitOp, err := config.clientResourceManager.Operations.Get(op.Name).Do()
+	if err != nil {
+		return fmt.Errorf("The folder '%s' has been created but we could not retrieve its id. Delete the folder manually and retry or use 'terraform import'.", displayName)
+	}
+	response := waitOp.Response.(map[string]interface{})
+	d.SetId(response["name"].(string))
+
+	return resourceGoogleFolderRead(d, meta)
+}
+
+func resourceGoogleFolderRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	folder, err := config.clientResourceManagerV2Beta1.Folders.Get(d.Id()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, d.Id())
+	}
+
+	d.Set("name", folder.Name)
+	d.Set("parent", folder.Parent)
+	d.Set("display_name", folder.DisplayName)
+	d.Set("lifecycle_state", folder.LifecycleState)
+	d.Set("create_time", folder.CreateTime)
+
+	return nil
+}
+
+func resourceGoogleFolderUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	displayName := d.Get("display_name").(string)
+
+	d.Partial(true)
+	if d.HasChange("display_name") {
+		_, err := config.clientResourceManagerV2Beta1.Folders.Patch(d.Id(), &resourceManagerV2Beta1.Folder{
+			DisplayName: displayName,
+		}).Do()
+
+		if err != nil {
+			return fmt.Errorf("Error updating display_name to '%s': %s", displayName, err)
+		}
+
+		d.SetPartial("display_name")
+	}
+
+	if d.HasChange("parent") {
+		newParent := d.Get("parent").(string)
+		op, err := config.clientResourceManagerV2Beta1.Folders.Move(d.Id(), &resourceManagerV2Beta1.MoveFolderRequest{
+			DestinationParent: newParent,
+		}).Do()
+
+		if err != nil {
+			return fmt.Errorf("Error moving folder '%s' to '%s': %s", displayName, newParent, err)
+		}
+
+		err = resourceManagerV2Beta1OperationWait(config, op, "move folder")
+		if err != nil {
+			return fmt.Errorf("Error moving folder '%s' to '%s': %s", displayName, newParent, err)
+		}
+
+		d.SetPartial("parent")
+	}
+
+	d.Partial(false)
+
+	return nil
+}
+
+func resourceGoogleFolderDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	displayName := d.Get("display_name").(string)
+
+	_, err := config.clientResourceManagerV2Beta1.Folders.Delete(d.Id()).Do()
+	if err != nil {
+		return fmt.Errorf("Error deleting folder %s", displayName)
+	}
+
```

Comment:
Missing a `d.SetId("")` call to indicate successful deletion

Although apparently that isn't needed since the tests passed? I'm curious if we should be doing that or not.

Reply:
In resource.go in terraform core:

```go
// Destroy the resource since it is created
if err := r.Delete(data, meta); err != nil {
  return r.recordCurrentSchemaVersion(data.State()), err
}

// Make sure the ID is gone.
data.SetId("")
```

It does it for us. Calling it again won't cause harm. Should I add it for consistency?

Reply:
Good to know, will omit that on delete in the future.